### PR TITLE
Fix SelectionTools not considering LineStyle

### LINKED ISSUE #3793: When changing the line style of something which is already drawn, if you already have the line style that you want selected, you must switch off the current line style and back to be able to change the items line style
(Please complete the following information, and then delete this line)

**Affects versions :**
 - OS: Arch Linux
 - (Linux only) Desktop environment: Gnome Wayland
 - Which version of libgtk do you use: 3.24.31
 - Version of Xournal++: 1.1.0
 - Installation method: Official package from the Arch Linux repositories

**Describe the bug**
If you have an item which is drawn with a specific line style, and you select the item to switch the line style, if you currently have the line style that you want the item to be switched to already seleced, and you try and select it again to switch the item, nothing happens. Instead you must switch off of the line style, select the item, then reselect the desired line style.

**To Reproduce**
Using a circle as an example
1. Select the pen tool
2. Select the standard line style
3. select the draw ellipse tool
4. Draw an ellipse
5. switch the pen to the dashed line line-style
6. select the ellipse
7. try and switch the ellipse to the dased line style by selecting the already selected dased line style, and observe that you cant switch the line style of the ellipse
8. switch back to the standard line style
9. reselect the ellipse
10. select the dased line style and observe that the ellipse is now using the dashed line style

**Expected behavior**
I expect that the selected item will change to whatever line style that I select regardless of what line style is currently selected.

### PR DESCRIPTION
**Pull request by forgedOctopus**
Fixes #3793.

Before the fix, the currently selected LineStyle of the Pen is marked as active in the toolbar during a selection. This prevents changing the LineStyle of the selection in certain situations:

https://user-images.githubusercontent.com/65341804/173325173-41a82a8d-d126-46f5-8fcc-158496136d97.mp4

The PR has the goal to be simple and make only few changes to the source code. It introduces a flag in the LineStyle class. The flag indicates whether the in the toolbar selected LineStyle should be cleared or not:
- if all selected stroke elements have the same LineStyle, the flag is not set and this Style will be marked as active in the toolbar
- if the selected stroke elements have different LineStyles, the flag is set and no LineStyle will be marked active in the toolbar
- if no elements are selected, the original behavior is kept

https://user-images.githubusercontent.com/65341804/173325164-e7210707-a0a0-45bb-80ab-ea1834ad0454.mp4

While the PR provides a simple fix, a larger re-write of the LineStyle implementation would probably make sense. An implementation using Enums like ToolSize's does, for instance, should make the LineStyle implementation more consistent with other Tool implementations and solve #2625.

If there is interest in such a re-write, I'm happy to look into it.

Looking forward to your feedback.

### CODE REVIEW

**Comment by forgedOctopus** [comment]
The PR should be ready for review again, sorry. 

I rebased the PR on the current master. For this, I did a force-push overwriting the PR's old history.

During the rebase, I made only one change:
I fixed the conflict in [705ac2c](https://github.com/xournalpp/xournalpp/pull/4089/commits/705ac2cd31f5e84789254545257dc529c5a236d6). In the master the import section of `src/core/gui/XournalView.cpp` had been updated. This resulted in a conflict with the by me in the PR added import of `model/StrokeStyle.h`.

Looking forward to your feedback.

**Comment by forgedOctopus** [comment]
Thanks for reaching out. No worries, I totally understand this.

In any case, feel free to ask if there's anything unclear or needs to be done.

**Comment by forgedOctopus** [comment]
> Looking forward to your feedback.

I realized, I might have formulated this badly (English is not my native language). To avoid misunderstandings: I didn't want to pressure you with this. I'm happy for any feedback, but I'm totally aware, that this can take time and this is not a matter of course. 

So no worries if your feedback takes some time.

File: src/core/control/Control.cpp [hunk 1]
```diff
@@ -1839,6 +1839,9 @@ void Control::toolChanged() {
         toolColorChanged();
     }
 
+    if (toolHandler->getToolType() == TOOL_PEN)
+        toolLineStyleChanged();
```
Comment (bhennion) [inline src/core/control/Control.cpp:1843]: Please always encapsulate conditional code with braces (see [our coding conventions](https://github.com/xournalpp/xournalpp/wiki/Coding-conventions#indentation))
```suggestion
    if (toolHandler->getToolType() == TOOL_PEN) {
        toolLineStyleChanged();
    }
```

File: src/core/model/LineStyle.h [hunk 2]
```diff
@@ -61,4 +70,9 @@ class LineStyle: public Serializable {
      * Dash count (0 for no dash)
      */
     int dashCount = 0;
+
+    /**
+     * Flag to clear LineStyle selection
+     */
+    bool clearLineStyle = false;
```
Comment (bhennion) [inline src/core/model/LineStyle.h:77]: I don't like this: the flag has nothing to do with the LineStyle itself, so it should not be part of the data structure.

I guess the way I'd do this is by adding a tool capability (see ToolEnums.h::ToolCapabilities and Tool::setCapability)

File: src/core/control/ToolHandler.cpp [hunk 3]
```diff
@@ -534,13 +534,18 @@ void ToolHandler::setSelectionEditTools(bool setColor, bool setSize, bool setFil
         t->setSize(TOOL_SIZE_NONE);
         t->setColor(Color(-1));
         t->setFill(false);
+
+        LineStyle newStyle = t->getLineStyle();
+        newStyle.setClearLineStyle(!setLineStyle);
+        t->setLineStyle(newStyle);
```
Comment (bhennion) [inline src/core/control/ToolHandler.cpp:540]:
```suggestion
        t->setCapability(TOOL_CAP_LINE_STYLE, setLineStyle);
        t->setSize(TOOL_SIZE_NONE);
        t->setColor(Color(-1));
        t->setFill(false);
```

with TOOL_CAP_LINE_STYLE declared in ToolEnums.h

File: src/core/model/LineStyle.cpp [hunk 4]
```diff
@@ -27,6 +27,8 @@ void LineStyle::operator=(const LineStyle& other) {
 
     other.getDashes(dashes, dashCount);
     setDashes(dashes, dashCount);
+
+    clearLineStyle = other.clearLineStyle;
```
Comment (bhennion) [inline src/core/model/LineStyle.cpp:31]:
```suggestion
```

File: src/core/model/LineStyle.cpp [hunk 5]
```diff
@@ -88,3 +90,8 @@ void LineStyle::setDashes(const double* dashes, int dashCount) {
  * @return true if dashed
  */
 auto LineStyle::hasDashes() const -> bool { return this->dashCount > 0; }
+
+
+void LineStyle::setClearLineStyle(bool clearLineStyle) { this->clearLineStyle = clearLineStyle; }
+
+bool LineStyle::getClearLineStyle() const { return clearLineStyle; }
```
Comment (bhennion) [inline src/core/model/LineStyle.cpp:97]:
```suggestion
```

File: src/core/model/LineStyle.h [hunk 6]
```diff
@@ -51,6 +51,15 @@ class LineStyle: public Serializable {
...
 
+    /**
+     * Set flag to clear the LineStyle selection in the toolbar
+     *
+     * @param clearLineStyle true, to clear toolbar LineStyle selection
+     */
+    void setClearLineStyle(bool clearLineStyle);
+
+    bool getClearLineStyle() const;
+
```
Comment (bhennion) [inline src/core/model/LineStyle.h:62]:
```suggestion
```

File: src/core/control/Control.cpp [hunk 7]
```diff
@@ -1976,9 +1979,27 @@ void Control::toolFillChanged() {
 }
 
 void Control::toolLineStyleChanged() {
+    if (toolHandler->getLineStyle().getClearLineStyle()) {
```
Comment (bhennion) [inline src/core/control/Control.cpp:1982]:
```suggestion
    if (!toolHandler->hasCapability(TOOL_CAP_LINE_STYLE)) {
```

**Review by bhennion — CHANGES_REQUESTED** [review]
First, thanks for this great PR!

There is one thing I dislike (see below). Otherwise, just nitpicks.

**Comment by forgedOctopus** [comment]
Thanks a lot for your review and helpful suggestions. 
I will start implementing the required changes.

File: src/core/control/Control.cpp (hunk 1 above)
Comment (forgedOctopus) [reply src/core/control/Control.cpp]: Fixed by 923338c2.

File: src/core/model/LineStyle.h (hunk 6 above)
Comment (forgedOctopus) [reply src/core/model/LineStyle.h]: Resolved by 76fab8d3808b620d2d7faeef98bfe48b56f03254.

File: src/core/model/LineStyle.cpp (hunk 5 above)
Comment (forgedOctopus) [reply src/core/model/LineStyle.cpp]: Resolved by 76fab8d3808b620d2d7faeef98bfe48b56f03254.

File: src/core/model/LineStyle.cpp (hunk 4 above)
Comment (forgedOctopus) [reply src/core/model/LineStyle.cpp]: Resolved by 76fab8d3808b620d2d7faeef98bfe48b56f03254.

File: src/core/control/ToolHandler.cpp (hunk 3 above)
Comment (forgedOctopus) [reply src/core/control/ToolHandler.cpp]: Resolved by baffc58cbe83b76317ee8a0c8f5497dd5b52c85b.

File: src/core/control/Control.cpp (hunk 7 above)
Comment (forgedOctopus) [reply src/core/control/Control.cpp]: Resolved by fea96d645547dc7db4ef2605aa2ad8c802c8b945 and 660d3f958c816c90568a3594831aa74ada84761c.

File: src/core/model/LineStyle.h (hunk 2 above)
Comment (forgedOctopus) [reply src/core/model/LineStyle.h]: I agree.

I decided to re-write the fix using a new tool capability as you suggested.
In my opinion, the new fix makes way more sense than the previous workaround and does not introduce code where it doesn't belong (e.g. in `XournalView::setSelection()`, `ToolHandler::setSelectionEditTools())`).

**Comment by forgedOctopus** [comment]
Based on your feedback, I decided to re-write the fix using a new tool capability for line style as you suggested.

I think the new fix makes way more sense now than the previous workaround. Furthermore, it improves the behavior: same as with tool size and color, the line style can now only be selected, when the tool actually supports the capability. I added the capability only for the pen tool. Hence, changing the line style is now only possible for the pen.

The video below illustrates the new behavior:

https://user-images.githubusercontent.com/65341804/180599610-d6bb27fb-6e29-41e9-bfa0-0a8fa45996d7.mp4

File: src/core/control/Control.cpp [hunk 8]
```diff
@@ -1979,7 +1992,35 @@ void Control::toolLineStyleChanged() {
...
 
-    if (style == "dash") {
+    bool isLineStyleSameForAll = true;
+	
+    if (win) {
+        EditSelection* sel = win->getXournal()->getSelection();
+        if (sel) {
+			bool isFirstStrokeElement = true;
+			string previous_style = "none";
+
```
Comment (bhennion) [inline src/core/control/Control.cpp:2002]: Just as a reminder, for when we pass to c++ 20

```suggestion
			
	\\ Todo(cpp20) Replace with std::ranges::filter_view and for_first_then_for_each
```

File: src/core/control/Control.cpp [hunk 9]
```diff
@@ -1979,7 +1992,35 @@ void Control::toolLineStyleChanged() {
...
+				if (e->getType() == ELEMENT_STROKE) {
+					auto* s = dynamic_cast<Stroke*>(e);
+					style = StrokeStyle::formatStyle(s->getLineStyle());
+
+					if (isFirstStrokeElement) {
+						previous_style = style;
+						isFirstStrokeElement = false;
+					} else {
+						if (style != previous_style) {
+							isLineStyleSameForAll = false;
```
Comment (bhennion) [inline src/core/control/Control.cpp:2013]:
```suggestion
							isLineStyleSameForAll = false;
							break;
```

File: src/core/control/Control.cpp [hunk 10]
```diff
@@ -1979,7 +1992,35 @@ void Control::toolLineStyleChanged() {
...
+    if (win) {
+        EditSelection* sel = win->getXournal()->getSelection();
+        if (sel) {
+			bool isFirstStrokeElement = true;
+			string previous_style = "none";
+			
+			for (Element* e: sel->getElements()) {
+				if (e->getType() == ELEMENT_STROKE) {
+					auto* s = dynamic_cast<Stroke*>(e);
+					style = StrokeStyle::formatStyle(s->getLineStyle());
```
Comment (bhennion) [inline src/core/control/Control.cpp:2006]: You may wanna filter with `s->getToolType() == STROKE_TOOL_PEN` as well here, to exclude highlighter or whiteout strokes.

**Review by bhennion — CHANGES_REQUESTED** [review]
Thanks for the changes! This is much cleaner indeed.

I still have a couple of comments (see below). Mainly: highlighter strokes should not count in "are there strokes of different line styles?".

I also have another request (it may be out of scope). See this video:

https://user-images.githubusercontent.com/17818041/180601196-01a7de7b-b1ca-4b88-bb13-5add4e67550b.mp4

When we change the line style of selected strokes to, say, dashes, it also changes the line style of the pen tool (so that strokes drawn after that are also dashed).

I'd rather the tool's line style was not affected by whatever happens to the selection (like for colors or stroke width).
This is not a bug introduced by your PR, so you can chose to declare it out of scope :-)

File: src/core/gui/XournalView.cpp [hunk 11]
```diff
@@ -637,16 +638,17 @@ void XournalView::setSelection(EditSelection* selection) {
             if (s->getToolType() != STROKE_TOOL_ERASER) {
                 canChangeColor = true;
                 canChangeFill = true;
+                canChangeLineStyle = true;
```
[the remark this answers is not in the document]
Comment (forgedOctopus) [reply src/core/gui/XournalView.cpp]: True, oversaw this one, thanks.

Resolved by 3c93210ce040a2e7074824f9dbd9b2e7d18eeeca.

File: src/core/control/Control.cpp (hunk 9 above)
Comment (forgedOctopus) [reply src/core/control/Control.cpp]: Resolved by 018fdef29d82d5a9ac214621cc5d6681ff840bad.

File: src/core/control/Control.cpp (hunk 10 above)
Comment (forgedOctopus) [reply src/core/control/Control.cpp]: Thanks, resolved by 67e5e9dea630e9ed617d84e44be6078f89b6ecee and the later refactorings.

File: src/core/control/Control.cpp (hunk 8 above)
Comment (forgedOctopus) [reply src/core/control/Control.cpp]: Added in a1f5563482acc2a3b6becf0801ac0b6a92f22363.

Sorry, due to the refactoring I'm not entirely sure if I put it in the correct spot.

Can you quickly check if it's correct?

Comment (bhennion) [reply src/core/control/Control.cpp]: Yes, that's the right place

**Comment by forgedOctopus** [comment]
Thanks for your quick review and comments!

I could resolve all comments with your suggestions.

I was also able to solve the issue with the tool line style being affected by the line style set for a selection (d4a9f69128e281b7326cf776179808bbcfae7e64).

In addition, I did some more code polish (f973f2689a844081c33b49df2e7a0feb345a0299, 018fdef29d82d5a9ac214621cc5d6681ff840bad, e498971daf98002a794e728600305b20f6da9cf5, ed3c18022169c0ed43615a668bf574b86bb20067).

I kept 018fdef29d82d5a9ac214621cc5d6681ff840bad and e498971daf98002a794e728600305b20f6da9cf5 both instead of one large refactoring commit, so you can see my thought process and decide which to keep / not to keep.  And if you don't think a certain refactoring was a good idea, I can easily revert it.

Below a video to illustrate the current behavior:

https://user-images.githubusercontent.com/65341804/180640363-eb8eac5a-bda8-48ba-8ad9-f4fb95245057.mp4

**Review by bhennion — APPROVED** [review]
LGTM.

If you're satisfied as well, we'll give other devs a couple of days to review if they want to, and then squash and merge.

@xournalpp/core Merging in 72 hours unless an objectino is raised.

**Comment by forgedOctopus** [comment]
Sounds good! :)
Thanks for taking your time and providing such helpful reviews.

**Pull request merged**